### PR TITLE
Async by default

### LINKED ISSUE #103: Verify that capturing lists works
```java
rest.get("/api").dispatch(series(),
        on(SUCCESSFUL).capture(Body.class, this::map))
        .to(listOf(String.class));

private List<String> map(final Body body) {
    // ...
}
```

### LINKED ISSUE #107: Check for consistent exception handling
- [ ] Are we catching/rethrowing/wrapping exceptions somewhere?
- [ ] If yes, should we?
- [ ] Define clear rules + tests for exception handling inside riptide itself
- [ ] Document it

### LINKED ISSUE #109: Make Riptide async by default?
Right now we always have to maintain two APIs: Rest (sync) and AsyncRest (async). Because of Riptide's design they both support the very same routing trees out of the box. This already makes Riptide stand out from the competition, since their APIs usually differ dramatically between the two approaches.

The problem we have right now is the different return value for async: `ListenableFuture<Capture>` vs. `Capture`. Translating from async to sync is relative easy:

```java
http.get(..)
        .dispatch(..)
        .to(Success.class);
```

vs.

```java
http.get(..)
        .dispatch(..)
        .get(10, SECONDS)
        .to(Success.class);
```

Given this, should we consider dropping the dedicated support for the synchronous version altogether?

### PR DESCRIPTION
**Pull request by whiskeysierra**
Fixes #109 
Fixes #107 
Fixes #103

### CODE REVIEW

File: src/main/java/org/zalando/riptide/RestClient.java [hunk 1]
```diff
@@ -20,133 +20,79 @@
...
-import org.springframework.web.client.RestClientException;
 
-import javax.annotation.Nullable;
-import java.io.IOException;
 import java.net.URI;
-import java.util.List;
-import java.util.Optional;
 
-import static java.lang.String.format;
+abstract class RestClient {
```
Comment (lukasniemeier-zalando) [inline src/main/java/org/zalando/riptide/RestClient.java]: Is this class still needed?

File: src/main/java/org/zalando/riptide/RestClient.java [hunk 2]
```diff
@@ -20,133 +20,79 @@
...
 
-abstract class RestClient<R> {
-
-    public final Requester<R> get(final String urlTemplate, final Object... urlVariables) {
+    public final Requester get(final String urlTemplate, final Object... urlVariables) {
         return execute(HttpMethod.GET, urlTemplate, urlVariables);
     }
 
-    public final Requester<R> get(final URI url) {
+    public final Requester get(final URI url) {
```
Comment (lukasniemeier-zalando) [inline src/main/java/org/zalando/riptide/RestClient.java]: IMO all those methods should be covered in #106 - clarifying the usage of the `UriTemplateHandler`

File: src/main/java/org/zalando/riptide/RestClient.java (hunk 1 above)
Comment (whiskeysierra) [inline src/main/java/org/zalando/riptide/RestClient.java]: No, it's not. Good catch.

**Comment by tkrop** [comment]
:+1:

File: src/main/java/org/zalando/riptide/Rest.java [hunk 3]
```diff
@@ -38,67 +43,205 @@
...
 
-        };
+            final Class<?> type = body.getClass();
+            @Nullable final MediaType contentType = headers.getContentType();
+
+            converters.stream()
+                    .filter(c -> c.canWrite(type, contentType))
+                    .map(this::<T>cast)
+                    .findFirst()
+                    .orElseThrow(() -> {
```
Comment (lukasniemeier-zalando) [inline src/main/java/org/zalando/riptide/Rest.java]: extract to method?

**Comment by lukasniemeier-zalando** [comment]
:+1:

**Pull request merged**